ngraph/frontend/paddlepaddle/src/model.cpp

```diff
@@ -502,6 +475,250 @@ namespace ngraph
+                    {
+                        FRONT_END_GENERAL_CHECK(std::find(target_out_tensor.begin(),
+                                                          target_out_tensor.end(),
+                                                          tensor) != target_out_tensor.end(),
+                                                "Incorrect Place for cutting.");
+                        new_outputs.push_back(tensor);
+                    }
+                }
+
+                std::swap(new_outputs, m_outputs);
+                m_graph_changed = true;
+                if (!new_name_optional.empty())
+                {
+                    setNameForTensor(place, new_name_optional);
+                }
+            }
+        }
+
+        void InputModelPDPD::InputModelPDPDImpl::cutAndAddNewInput(
+            Place::Ptr place, const std::string& new_name_optional)
+        {
+            auto tensor_place = pdpd::castToTensorPlace(place);
+            if (tensor_place)
+            {
+                m_inputs.push_back(tensor_place);
+                std::vector<std::shared_ptr<TensorPlacePDPD>> new_tensors;
+                traverse_up(m_outputs, nullptr, &new_tensors);
+
+                std::vector<Place::Ptr> new_inputs;
+                for (const auto& in : m_inputs)
+                {
+                    if (std::find(new_tensors.begin(), new_tensors.end(), in) != new_tensors.end())
+                        new_inputs.push_back(in);
+                }
+                std::swap(m_inputs, new_inputs);
+                m_graph_changed = true;
+                if (!new_name_optional.empty())
+                {
+                    setNameForTensor(place, new_name_optional);
+                }
+            }
+        }
+
+        void InputModelPDPD::InputModelPDPDImpl::freeNameForTensor(const std::string& name)
+        {
+            if (m_var_names.count(name) != 0)
+            {
+                m_var_names[name]->remove_name(name);
+                m_var_names.erase(name);
+            }
+        }
+
+        void InputModelPDPD::InputModelPDPDImpl::addNameForTensor(Place::Ptr tensor,
+                                                                  const std::string& new_name)
+        {
+            auto it = std::find(m_var_places.begin(), m_var_places.end(), tensor);
+            FRONT_END_GENERAL_CHECK(it != m_var_places.end(),
+                                    "Model doesn't own the provided tensor.");
+            FRONT_END_GENERAL_CHECK(m_var_names.count(new_name) == 0,
+                                    "The provided name is already used in the model");
+            (*it)->add_name(new_name);
+            m_var_names[new_name] = *it;
+        }
+
+        void InputModelPDPD::InputModelPDPDImpl::setNameForTensor(Place::Ptr tensor,
+                                                                  const std::string& new_name)
+        {
+            auto it = std::find(m_var_places.begin(), m_var_places.end(), tensor);
```

Comment:
We're searching for place in the model, which has linear complexity
Instead, can 'Place' just check that it's 'm_inputModel' is the same as 'this'?

Reply:
I think so, good idea, thanks

Reply:
done